File: app/src/main/java/awais/instagrabber/webservices/StoriesService.java
package awais.instagrabber.webservices;

import android.util.Log;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;
import java.util.UUID;

import awais.instagrabber.models.FeedStoryModel;
import awais.instagrabber.models.HighlightModel;
import awais.instagrabber.models.ProfileModel;
import awais.instagrabber.models.StoryModel;
import awais.instagrabber.repositories.StoriesRepository;
import awais.instagrabber.repositories.responses.StoryStickerResponse;
import awais.instagrabber.utils.Constants;
import awais.instagrabber.utils.ResponseBodyUtils;
import awais.instagrabber.utils.TextUtils;
import awais.instagrabber.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class StoriesService extends BaseService {
    private static final String TAG = "StoriesService";
    private static final boolean loadFromMock = false;

    private final StoriesRepository repository;

    private static StoriesService instance;

    private StoriesService() {
        final Retrofit retrofit = getRetrofitBuilder()
                .baseUrl("https://i.instagram.com")
                .build();
        repository = retrofit.create(StoriesRepository.class);
    }

    public static StoriesService getInstance() {
        if (instance == null) {
            instance = new StoriesService();
        }
        return instance;
    }

    public void fetch(final String mediaId,
                      final ServiceCallback<StoryModel> callback) {
        final Call<String> request = repository.fetch(mediaId);
        request.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull final Call<String> call,
                                   @NonNull final Response<String> response) {
                if (callback == null) return;
                final String body = response.body();
                if (body == null) {
                    callback.onSuccess(null);
                    return;
                }
                try {
                    final JSONObject itemJson = new JSONObject(body).getJSONArray("items").getJSONObject(0);
                    callback.onSuccess(ResponseBodyUtils.parseStoryItem(itemJson, false, false, null));
                } catch (JSONException e) {
                    callback.onFailure(e);
                }
            }

            @Override
            public void onFailure(@NonNull final Call<String> call,
                                  @NonNull final Throwable t) {
                if (callback != null) {
                    callback.onFailure(t);
                }
            }
        });
    }

    public void getFeedStories(final String csrfToken, final ServiceCallback<List<FeedStoryModel>> callback) {
        final Map<String, Object> form = new HashMap<>(4);
        form.put("reason", "cold_start");
        form.put("_csrftoken", csrfToken);
        form.put("_uuid", UUID.randomUUID().toString());
        form.put("supported_capabilities_new", Constants.SUPPORTED_CAPABILITIES);
        final Map<String, String> signedForm = Utils.sign(form);
        final Call<String> response = repository.getFeedStories(Constants.I_USER_AGENT, signedForm);
        response.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull final Call<String> call, @NonNull final Response<String> response) {
                final String body = response.body();
                if (body == null) {
                    Log.e(TAG, "getFeedStories: body is empty");
                    return;
                }
                parseStoriesBody(body, callback);
            }

            @Override
            public void onFailure(@NonNull final Call<String> call, @NonNull final Throwable t) {
                callback.onFailure(t);
            }
        });
    }

    private void parseStoriesBody(final String body, final ServiceCallback<List<FeedStoryModel>> callback) {
        try {
            final List<FeedStoryModel> feedStoryModels = new ArrayList<>();
            final JSONArray feedStoriesReel = new JSONObject(body).getJSONArray("tray");
            for (int i = 0; i < feedStoriesReel.length(); ++i) {
                final JSONObject node = feedStoriesReel.getJSONObject(i);
                final JSONObject user = node.getJSONObject(node.has("user") ? "user" : "owner");
                final ProfileModel profileModel = new ProfileModel(false, false, false,
                                                                   user.getString("pk"),
                                                                   user.getString("username"),
                                                                   null, null, null,
                                                                   user.getString("profile_pic_url"),
                                                                   null, 0, 0, 0, false, false, false, false, false);
                final String id = node.getString("id");
                final long timestamp = node.getLong("latest_reel_media");
                final boolean fullyRead = !node.isNull("seen") && node.getLong("seen") == timestamp;
                final JSONObject itemJson = node.getJSONArray("items").getJSONObject(0);
                final StoryModel firstStoryModel = ResponseBodyUtils.parseStoryItem(itemJson, false, false, null);
                feedStoryModels.add(new FeedStoryModel(id, profileModel, fullyRead, timestamp, firstStoryModel));
            }
            callback.onSuccess(feedStoryModels);
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing json", e);
        }
    }

    public void fetchHighlights(final String profileId,
                                final ServiceCallback<List<HighlightModel>> callback) {
        final Call<String> request = repository.fetchHighlights(profileId);
        request.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull final Call<String> call, @NonNull final Response<String> response) {
                try {
                    if (callback == null) {
                        return;
                    }
                    final String body = response.body();
                    if (TextUtils.isEmpty(body)) {
                        callback.onSuccess(null);
                        return;
                    }
                    final JSONArray highlightsReel = new JSONObject(body).getJSONArray("tray");

                    final int length = highlightsReel.length();
                    final List<HighlightModel> highlightModels = new ArrayList<>();

                    for (int i = 0; i < length; ++i) {
                        final JSONObject highlightNode = highlightsReel.getJSONObject(i);
                        highlightModels.add(new HighlightModel(
                                highlightNode.getString("title"),
                                highlightNode.getString(Constants.EXTRAS_ID),
                                highlightNode.getJSONObject("cover_media")
                                        .getJSONObject("cropped_image_version")
                                        .getString("url"),
                                highlightNode.getLong("latest_reel_media")
                        ));
                    }
                    callback.onSuccess(highlightModels);
                } catch (JSONException e) {
                    Log.e(TAG, "onResponse", e);
                    callback.onFailure(e);
                }
            }

            @Override
            public void onFailure(@NonNull final Call<String> call, @NonNull final Throwable t) {
                if (callback != null) {
                    callback.onFailure(t);
                }
            }
        });
    }

    public void fetchArchive(final String maxId,
                             final ServiceCallback<ArchiveFetchResponse> callback) {
        final Map<String, String> form = new HashMap<>();
        form.put("include_suggested_highlights", "false");
        form.put("is_in_archive_home", "true");
        form.put("include_cover", "1");
        form.put("timezone_offset", String.valueOf(TimeZone.getDefault().getRawOffset() / 1000));
        if (!TextUtils.isEmpty(maxId)) {
            form.put("max_id", maxId); // NOT TESTED
        }
        final Call<String> request = repository.fetchArchive(form);
        request.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull final Call<String> call, @NonNull final Response<String> response) {
                try {
                    if (callback == null) {
                        return;
                    }
                    final String body = response.body();
                    if (TextUtils.isEmpty(body)) {
                        callback.onSuccess(null);
                        return;
                    }
                    final JSONObject data = new JSONObject(body);
                    final JSONArray highlightsReel = data.getJSONArray("items");

                    final int length = highlightsReel.length();
                    final List<HighlightModel> highlightModels = new ArrayList<>();

                    for (int i = 0; i < length; ++i) {
                        final JSONObject highlightNode = highlightsReel.getJSONObject(i);
                        highlightModels.add(new HighlightModel(
                                null,
                                highlightNode.getString(Constants.EXTRAS_ID),
                                highlightNode.getJSONObject("cover_image_version").getString("url"),
                                highlightNode.getLong("timestamp")
                        ));
                    }
                    callback.onSuccess(new ArchiveFetchResponse(highlightModels,
                                                                data.getBoolean("more_available"),
                                                                data.getString("max_id")));
                } catch (JSONException e) {
                    Log.e(TAG, "onResponse", e);
                    callback.onFailure(e);
                }
            }

            @Override
            public void onFailure(@NonNull final Call<String> call, @NonNull final Throwable t) {
                if (callback != null) {
                    callback.onFailure(t);
                }
            }
        });
    }

    public void getUserStory(final String id,
                             final String username,
                             final boolean isLoc,
                             final boolean isHashtag,
                             final boolean highlight,
                             final ServiceCallback<List<StoryModel>> callback) {
        final String url = buildUrl(id, isLoc, isHashtag, highlight);
        final Call<String> userStoryCall = repository.getUserStory(Constants.I_USER_AGENT, url);
        userStoryCall.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull final Call<String> call, @NonNull final Response<String> response) {
                JSONObject data;
                String localUsername = username;
                try {
                    final String body = response.body();
                    if (body == null) {
                        Log.e(TAG, "body is null");
                        return;
                    }
                    data = new JSONObject(body);

                    if (!highlight)
                        data = data.optJSONObject((isLoc || isHashtag) ? "story" : "reel");
                    else if (highlight) data = data.getJSONObject("reels").optJSONObject(id);

                    if (data != null
                            && localUsername == null
                            && !isLoc
                            && !isHashtag)
                        localUsername = data.getJSONObject("user").getString("username");

                    JSONArray media;
                    if (data != null
                            && (media = data.optJSONArray("items")) != null
                            && media.length() > 0 && media.optJSONObject(0) != null) {

                        final int mediaLen = media.length();
                        final List<StoryModel> models = new ArrayList<>();
                        for (int i = 0; i < mediaLen; ++i) {
                            data = media.getJSONObject(i);
                            models.add(ResponseBodyUtils.parseStoryItem(data, isLoc, isHashtag, localUsername));
                        }
                        callback.onSuccess(models);
                    } else {
                        callback.onSuccess(null);
                    }
                } catch (JSONException e) {
                    Log.e(TAG, "Error parsing string");
                }
            }

            @Override
            public void onFailure(@NonNull final Call<String> call, @NonNull final Throwable t) {
                callback.onFailure(t);
            }
        });
    }

    private void respondToSticker(final String storyId,
                                  final String stickerId,
                                  final String action,
                                  final String arg1,
                                  final String arg2,
                                  final String userId,
                                  final String csrfToken,
                                  final ServiceCallback<StoryStickerResponse> callback) {
        final Map<String, Object> form = new HashMap<>();
        form.put("_csrftoken", csrfToken);
        form.put("_uid", userId);
        form.put("_uuid", UUID.randomUUID().toString());
        form.put("mutation_token", UUID.randomUUID().toString());
        form.put("client_context", UUID.randomUUID().toString());
        form.put("radio_type", "wifi-none");
        form.put(arg1, arg2);
        final Map<String, String> signedForm = Utils.sign(form);
        final Call<StoryStickerResponse> request =
                repository.respondToSticker(Constants.I_USER_AGENT, storyId, stickerId, action, signedForm);
        request.enqueue(new Callback<StoryStickerResponse>() {
            @Override
            public void onResponse(@NonNull final Call<StoryStickerResponse> call,
                                   @NonNull final Response<StoryStickerResponse> response) {
                if (callback != null) {
                    callback.onSuccess(response.body());
                }
            }

            @Override
            public void onFailure(@NonNull final Call<StoryStickerResponse> call,
                                  @NonNull final Throwable t) {
                if (callback != null) {
                    callback.onFailure(t);
                }
            }
        });
    }

    // RespondAction.java
    public void respondToQuestion(final String storyId,
                                   final String stickerId,
                                   final String answer,
                                   final String userId,
                                   final String csrfToken,
                                   final ServiceCallback<StoryStickerResponse> callback) {
        respondToSticker(storyId, stickerId, "story_question_response", "response", answer, userId, csrfToken, callback);
    }

    // QuizAction.java
    public void respondToQuiz(final String storyId,
                               final String stickerId,
                               final int answer,
                               final String userId,
                               final String csrfToken,
                               final ServiceCallback<StoryStickerResponse> callback) {
        respondToSticker(storyId, stickerId, "story_quiz_answer", "answer", String.valueOf(answer), userId, csrfToken, callback);
    }

    // VoteAction.java
    public void respondToPoll(final String storyId,
                               final String stickerId,
                               final int answer,
                               final String userId,
                               final String csrfToken,
                               final ServiceCallback<StoryStickerResponse> callback) {
        respondToSticker(storyId, stickerId, "story_poll_vote", "vote", String.valueOf(answer), userId, csrfToken, callback);
    }

    public void respondToSlider(final String storyId,
                              final String stickerId,
                              final double answer,
                              final String userId,
                              final String csrfToken,
                              final ServiceCallback<StoryStickerResponse> callback) {
        respondToSticker(storyId, stickerId, "story_slider_vote", "vote", String.valueOf(answer), userId, csrfToken, callback);
    }

    private String buildUrl(final String id, final boolean isLoc, final boolean isHashtag, final boolean highlight) {
        final String userId = id.replace(":", "%3A");
        final StringBuilder builder = new StringBuilder();
        builder.append("https://i.instagram.com/api/v1/");
        if (isLoc) {
            builder.append("locations/");
        } else if (isHashtag) {
            builder.append("tags/");
        } else if (highlight) {
            builder.append("feed/reels_media/?user_ids=");
        } else {
            builder.append("feed/user/");
        }
        builder.append(userId);
        if (!highlight) {
            builder.append("/story/");
        }
        return builder.toString();
    }

    public class ArchiveFetchResponse {
        private final List<HighlightModel> archives;
        private final boolean hasNextPage;
        private final String nextCursor;

        public ArchiveFetchResponse(final List<HighlightModel> archives, final boolean hasNextPage, final String nextCursor) {
            this.archives = archives;
            this.hasNextPage = hasNextPage;
            this.nextCursor = nextCursor;
        }

        public List<HighlightModel> getArchives() {
            return archives;
        }

        public boolean hasNextPage() {
            return hasNextPage;
        }

        public String getNextCursor() {
            return nextCursor;
        }
    }

}

File: app/src/main/java/awais/instagrabber/webservices/StoriesService.kt
package awais.instagrabber.webservices

import android.util.Log
import awais.instagrabber.fragments.settings.PreferenceKeys
import awais.instagrabber.models.FeedStoryModel
import awais.instagrabber.models.HighlightModel
import awais.instagrabber.models.StoryModel
import awais.instagrabber.repositories.StoriesRepository
import awais.instagrabber.repositories.requests.StoryViewerOptions
import awais.instagrabber.repositories.responses.StoryStickerResponse
import awais.instagrabber.repositories.responses.User
import awais.instagrabber.utils.Constants
import awais.instagrabber.utils.ResponseBodyUtils
import awais.instagrabber.utils.TextUtils.isEmpty
import awais.instagrabber.utils.Utils
import awais.instagrabber.utils.extensions.TAG
import awais.instagrabber.webservices.RetrofitFactory.retrofit
import org.json.JSONArray
import org.json.JSONObject
import java.util.*

object StoriesService {
    private val repository: StoriesRepository = retrofit.create(StoriesRepository::class.java)

    suspend fun fetch(mediaId: Long): StoryModel {
        val response = repository.fetch(mediaId)
        val itemJson = JSONObject(response).getJSONArray("items").getJSONObject(0)
        return ResponseBodyUtils.parseStoryItem(itemJson, false, null)
    }

    suspend fun getFeedStories(): List<FeedStoryModel> {
        val response = repository.getFeedStories()
        return parseStoriesBody(response)
    }

    private fun parseStoriesBody(body: String): List<FeedStoryModel> {
        val feedStoryModels: MutableList<FeedStoryModel> = ArrayList()
        val feedStoriesReel = JSONObject(body).getJSONArray("tray")
        for (i in 0 until feedStoriesReel.length()) {
            val node = feedStoriesReel.getJSONObject(i)
            if (node.optBoolean("hide_from_feed_unit") && Utils.settingsHelper.getBoolean(PreferenceKeys.HIDE_MUTED_REELS)) continue
            val userJson = node.getJSONObject(if (node.has("user")) "user" else "owner")
            try {
                val user = User(userJson.getLong("pk"),
                    userJson.getString("username"),
                    userJson.optString("full_name"),
                    userJson.optBoolean("is_private"),
                    userJson.getString("profile_pic_url"),
                    userJson.optBoolean("is_verified")
                )
                val timestamp = node.getLong("latest_reel_media")
                val fullyRead = !node.isNull("seen") && node.getLong("seen") == timestamp
                val itemJson = if (node.has("items")) node.getJSONArray("items").optJSONObject(0) else null
                var firstStoryModel: StoryModel? = null
                if (itemJson != null) {
                    firstStoryModel = ResponseBodyUtils.parseStoryItem(itemJson, false, null)
                }
                feedStoryModels.add(FeedStoryModel(
                    node.getString("id"),
                    user,
                    fullyRead,
                    timestamp,
                    firstStoryModel,
                    node.getInt("media_count"),
                    false,
                    node.optBoolean("has_besties_media")))
            } catch (e: Exception) {
                Log.e(TAG, "parseStoriesBody: ", e)
            } // to cover promotional reels with non-long user pk's
        }
        val broadcasts = JSONObject(body).getJSONArray("broadcasts")
        for (i in 0 until broadcasts.length()) {
            val node = broadcasts.getJSONObject(i)
            val userJson = node.getJSONObject("broadcast_owner")
            val user = User(userJson.getLong("pk"),
                userJson.getString("username"),
                userJson.optString("full_name"),
                userJson.optBoolean("is_private"),
                userJson.getString("profile_pic_url"),
                userJson.optBoolean("is_verified")
            )
            feedStoryModels.add(FeedStoryModel(
                node.getString("id"),
                user,
                false,
                node.getLong("published_time"),
                ResponseBodyUtils.parseBroadcastItem(node),
                1,
                isLive = true,
                isBestie = false
            ))
        }
        return sort(feedStoryModels)
    }

    suspend fun fetchHighlights(profileId: Long): List<HighlightModel> {
        val response = repository.fetchHighlights(profileId)
        val highlightsReel = JSONObject(response).getJSONArray("tray")
        val length = highlightsReel.length()
        val highlightModels: MutableList<HighlightModel> = ArrayList()
        for (i in 0 until length) {
            val highlightNode = highlightsReel.getJSONObject(i)
            highlightModels.add(HighlightModel(
                highlightNode.getString("title"),
                highlightNode.getString(Constants.EXTRAS_ID),
                highlightNode.getJSONObject("cover_media")
                    .getJSONObject("cropped_image_version")
                    .getString("url"),
                highlightNode.getLong("latest_reel_media"),
                highlightNode.getInt("media_count")
            ))
        }
        return highlightModels
    }

    suspend fun fetchArchive(maxId: String): ArchiveFetchResponse {
        val form = mutableMapOf(
            "include_suggested_highlights" to "false",
            "is_in_archive_home" to "true",
            "include_cover" to "1",
        )
        if (!isEmpty(maxId)) {
            form["max_id"] = maxId // NOT TESTED
        }
        val response = repository.fetchArchive(form)
        val data = JSONObject(response)
        val highlightsReel = data.getJSONArray("items")
        val length = highlightsReel.length()
        val highlightModels: MutableList<HighlightModel> = ArrayList()
        for (i in 0 until length) {
            val highlightNode = highlightsReel.getJSONObject(i)
            highlightModels.add(HighlightModel(
                null,
                highlightNode.getString(Constants.EXTRAS_ID),
                highlightNode.getJSONObject("cover_image_version").getString("url"),
                highlightNode.getLong("latest_reel_media"),
                highlightNode.getInt("media_count")
            ))
        }
        return ArchiveFetchResponse(highlightModels, data.getBoolean("more_available"), data.getString("max_id"))
    }

    suspend fun getUserStory(options: StoryViewerOptions): List<StoryModel> {
        val url = buildUrl(options) ?: return emptyList()
        val response = repository.getUserStory(url)
        val isLocOrHashtag = options.type == StoryViewerOptions.Type.LOCATION || options.type == StoryViewerOptions.Type.HASHTAG
        val isHighlight = options.type == StoryViewerOptions.Type.HIGHLIGHT || options.type == StoryViewerOptions.Type.STORY_ARCHIVE
        var data: JSONObject? = JSONObject(response)
        data = if (!isHighlight) {
            data?.optJSONObject(if (isLocOrHashtag) "story" else "reel")
        } else {
            data?.getJSONObject("reels")?.optJSONObject(options.name)
        }
        var username: String? = null
        if (data != null && !isLocOrHashtag) {
            username = data.getJSONObject("user").getString("username")
        }
        val media: JSONArray? = data?.optJSONArray("items")
        return if (media?.length() ?: 0 > 0 && media?.optJSONObject(0) != null) {
            val mediaLen = media.length()
            val models: MutableList<StoryModel> = ArrayList()
            for (i in 0 until mediaLen) {
                data = media.getJSONObject(i)
                models.add(ResponseBodyUtils.parseStoryItem(data, isLocOrHashtag, username))
            }
            models
        } else emptyList()
    }

    private suspend fun respondToSticker(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        storyId: String,
        stickerId: String,
        action: String,
        arg1: String,
        arg2: String,
    ): StoryStickerResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uid" to userId,
            "_uuid" to deviceUuid,
            "mutation_token" to UUID.randomUUID().toString(),
            "client_context" to UUID.randomUUID().toString(),
            "radio_type" to "wifi-none",
            arg1 to arg2,
        )
        val signedForm = Utils.sign(form)
        return repository.respondToSticker(storyId, stickerId, action, signedForm)
    }

    suspend fun respondToQuestion(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        storyId: String,
        stickerId: String,
        answer: String,
    ): StoryStickerResponse = respondToSticker(csrfToken, userId, deviceUuid, storyId, stickerId, "story_question_response", "response", answer)

    suspend fun respondToQuiz(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        storyId: String,
        stickerId: String,
        answer: Int,
    ): StoryStickerResponse {
        return respondToSticker(csrfToken, userId, deviceUuid, storyId, stickerId, "story_quiz_answer", "answer", answer.toString())
    }

    suspend fun respondToPoll(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        storyId: String,
        stickerId: String,
        answer: Int,
    ): StoryStickerResponse = respondToSticker(csrfToken, userId, deviceUuid, storyId, stickerId, "story_poll_vote", "vote", answer.toString())

    suspend fun respondToSlider(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        storyId: String,
        stickerId: String,
        answer: Double,
    ): StoryStickerResponse = respondToSticker(csrfToken, userId, deviceUuid, storyId, stickerId, "story_slider_vote", "vote", answer.toString())

    suspend fun seen(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        storyMediaId: String,
        takenAt: Long,
        seenAt: Long,
    ): String {
        val reelsForm = mapOf(storyMediaId to listOf(takenAt.toString() + "_" + seenAt))
        val form = mutableMapOf(
            "_csrftoken" to csrfToken,
            "_uid" to userId,
            "_uuid" to deviceUuid,
            "container_module" to "feed_timeline",
            "reels" to reelsForm,
        )
        val signedForm = Utils.sign(form)
        val queryMap = mapOf(
            "reel" to "1",
            "live_vod" to "0",
        )
        return repository.seen(queryMap, signedForm)
    }

    private fun buildUrl(options: StoryViewerOptions): String? {
        val builder = StringBuilder()
        builder.append("https://i.instagram.com/api/v1/")
        val type = options.type
        var id: String? = null
        when (type) {
            StoryViewerOptions.Type.HASHTAG -> {
                builder.append("tags/")
                id = options.name
            }
            StoryViewerOptions.Type.LOCATION -> {
                builder.append("locations/")
                id = options.id.toString()
            }
            StoryViewerOptions.Type.USER -> {
                builder.append("feed/user/")
                id = options.id.toString()
            }
            StoryViewerOptions.Type.HIGHLIGHT, StoryViewerOptions.Type.STORY_ARCHIVE -> {
                builder.append("feed/reels_media/?user_ids=")
                id = options.name
            }
            StoryViewerOptions.Type.STORY -> {
            }
            else -> {
            }
        }
        if (id == null) {
            return null
        }
        builder.append(id)
        if (type != StoryViewerOptions.Type.HIGHLIGHT && type != StoryViewerOptions.Type.STORY_ARCHIVE) {
            builder.append("/story/")
        }
        return builder.toString()
    }

    private fun sort(list: List<FeedStoryModel>): List<FeedStoryModel> {
        val listCopy = ArrayList(list)
        listCopy.sortWith { o1, o2 ->
            when (Utils.settingsHelper.getString(PreferenceKeys.STORY_SORT)) {
                "1" -> return@sortWith o2.timestamp.compareTo(o1.timestamp)
                "2" -> return@sortWith o1.timestamp.compareTo(o2.timestamp)
                else -> return@sortWith 0
            }
        }
        return listCopy
    }

    class ArchiveFetchResponse(val result: List<HighlightModel>, val hasNextPage: Boolean, val nextCursor: String) {
        fun hasNextPage(): Boolean {
            return hasNextPage
        }
    }
}